extensions/amp-ad-network-fake-impl/0.1/amp-ad-network-fake-impl.js

```diff
@@ -106,55 +112,191 @@ export class AmpAdNetworkFakeImpl extends AmpA4A {
+      const img = imgs[i];
+      let width,
+        height,
+        area = -1;
+      if (img.hasAttribute('width')) {
+        width = img.getAttribute('width');
+      }
+      if (img.hasAttribute('height')) {
+        height = img.getAttribute('height');
+      }
+      if (height && width) {
+        area = height * width;
+      }
+      imgMetadata.push({
+        src: img.getAttribute('src'),
+        area,
+      });
     }
 
-    let creative = root./*OK*/ outerHTML;
-
-    // Metadata
-    creative += '<script type="application/json" amp-ad-metadata>';
-    creative += '{';
-
-    if (this.element.hasAttribute('amp-story')) {
-      const ctaTypeMeta = root.querySelector('[name=amp-cta-type]');
-      const ctaType = ctaTypeMeta && ctaTypeMeta.content;
-      creative += `"ctaType": "${ctaType}", `;
+    const json = doc.querySelectorAll('script[type]');
+    for (let i = 0; i < json.length; i++) {
+      const script = json[i];
+      const type = script.getAttribute('type');
+      if (
+        type == 'application/json' &&
+        script.hasAttribute('id') &&
+        !jsonMetadata.includes(script.getAttribute('id'))
+      ) {
+        jsonMetadata.push(script.getAttribute('id'));
+      }
+      if (
+        type == 'application/json' &&
+        script.hasAttribute('amp-ad-metadata')
+      ) {
+        const parsed = parseJson(script.textContent);
+        for (const attribute in parsed) {
+          metadata[attribute] = parsed[attribute];
+        }
+      }
+    }
+    const ampAnalytics = doc.querySelector('amp-analytics');
+    if (ampAnalytics && !jsonMetadata.includes('amp-analytics')) {
+      jsonMetadata.push('amp-analytics');
+    }
 
-      const ctaTypeUrl = root.querySelector('[name=amp-cta-url]');
-      const ctaUrl = ctaTypeUrl && ctaTypeUrl.content;
-      creative += `"ctaUrl": "${ctaUrl}", `;
+    const creative = doc.documentElement./*REVIEW*/ outerHTML;
+    let start = 0;
+    let end = 0;
+    if (firstRuntimeElement != null) {
+      firstRuntimeElement = firstRuntimeElement./*REVIEW*/ outerHTML;
+      lastRuntimeElement = lastRuntimeElement./*REVIEW*/ outerHTML;
+      start = creative.indexOf(firstRuntimeElement);
+      end = creative.indexOf(lastRuntimeElement) + lastRuntimeElement.length;
     }
+    metadata['ampRuntimeUtf16CharOffsets'] = [start, end];
 
-    creative += '"ampRuntimeUtf16CharOffsets": [0, 0],';
-    creative += '"customElementExtensions": [';
-    for (let i = 0; i < extensions.length; i++) {
-      if (i > 0) {
-        creative += ',';
+    if (jsonMetadata.length > 0) {
+      metadata['jsonUtf16CharOffsets'] = {};
+      for (let i = 0; i < jsonMetadata.length; i++) {
+        const name = jsonMetadata[i];
+        let nameElementString;
+        if (name != 'amp-analytics') {
+          nameElementString = doc.getElementById(name)./*REVIEW*/ outerHTML;
+        } else {
+          nameElementString = doc.querySelector('amp-analytics').childNodes[0]
```

Comment:
The integration test bug is here. The first child node is a textNode. Also we already did the querySelector above.